rfcs/0024-hard-forks.md

```diff
@@ -0,0 +1,455 @@
+this process is very conservative. The timeline will be as follows:
+
+1. Beginning of previous era:
+  * Discussion, design and implementation of features for next era begins. This
+    can happen in parallel for different features, and should be mostly done
+    before the finalization process begins.
+1. Six weeks from era start:
+  * A public testnet starts, forked off the mainnet at the current block. At
+    this point it has no changes from the previous era, other than being
+    segregated. This will be used as a dry-run for the upcoming fork. We may
+    need to modify delegations in order to ensure sufficient stake is online.
+  * Feature specification lock. All features proposed to go into this hard fork
+    have been submitted to the RFC process. Everything submitted must have a
+    working, tested implementation.
+2. Five weeks from era start:
+  * All proposals are accepted or rejected/pushed to the next era.
+3. Four weeks from era start:
+  * Testnet era advances and the new features begin the final burn-in test.
+    Builds should be available implementing everything, at a "release candidate"
+    level of quality.
+  * The Coda team and general public monitor the testnet, use it, and try to
+    break it.
+4. Up until one week from era start:
+  * For every issue we discover, if fixing it would require another hard fork we
+    disable the associated feature unless the change is very simple. Disabling a
+    feature requires starting a new era on the testnet. If fixing it is only a
+    code change we're a little more lenient: if the fix is minor we make it and
+    continue, if it's major we still drop the feature. In extreme situations,
+    where disabling an individual feature is not possible or advisable, we may
+    abort the upgrade entirely, releasing new clients that extend the current
+    era up to the next date.
+5. One week from era start:
+  * Mainnet client builds that support the new era are released and all nodes
+    are expected to upgrade. If no further intervention occurs the transition
+    will proceed as planned.
+  * No further changes are allowed. If a bug that should block era transition is
+    found we delay the era start by another week, releasing new clients, and
+    return to step 4.
+6. Era start:
+  * All changes are active on the mainnet and the burn-in testnet is retired.
+    Pop the confetti.
+
+For emergency fixes the procedure is much shorter. Anything that threatens
+security of funds, network liveness, or consistency counts as an emergency. We
+could also do emergency forks to revert thefts or loss of funds due to error,
+as in Ethereum's [DAO fork](https://eips.ethereum.org/EIPS/eip-779), but whether
+and when to do that is outside the scope of this RFC.
+
+Timeline:
+1. The underlying issue is discovered and documented. The Coda team decides an
+   emergency hard fork is necessary.
+2. We schedule the new era, for some time in the near future e.g. 48 hours from
```

Comment:
To limit centralization, other protocols often include some sort of "signaling" mechanism. This has the advantage that a client can:

- realize the network is trying to update without reviewing a centralized source
- a node can see whether the update is broadly accepted by the network in a decentralized way